Runner compatibility

jacksontj opened the pull request:
I _believe_ this is the last of it, covers the cases mentioned in #19707, and should finally make @cachedout 's runner stuff backwards compatible ;)

jacksontj committed:
Make non-async runner executions happen in a thread instead of a daemonized process.

This means we are backwards compatible with builtin print (yay!) and that you can control-c out of the execution (since it will kill the thread too).

jacksontj committed:
Inject globals into *all* runners' global namespace

Fix for #19707
Similar fix to #19712, just doing it once per mod.

jacksontj committed:
Remove unused import

cachedout commented:
Sweet! Thank you sir.

jacksontj commented:
I wonder if it'd be better to add that daemonize flag to async and move the threading/multiprocessing down there as well...

cachedout commented:
Seems like a more natural place to me. That might be worth moving. My $0.02 at least.

jacksontj commented:
Yea, I think I'll hold off on that move for a bit though. Have to think of
what flags we want to expose in that api.
On Jan 14, 2015 8:25 AM, "Mike Place" <notifications@github.com> wrote:

> Seems like a more natural place to me. That might be worth moving. My
> $0.02 at least.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/19729#issuecomment-69943297>.
>

cachedout commented:
Sounds good. I'm going to pull down these changes and check 'em out while Jenkins does its thing.

cachedout commented:
Would you mind grabbing the lint errors when you get a moment?

jacksontj commented:
Yea, I'll grab those
On Jan 14, 2015 8:29 AM, "Mike Place" <notifications@github.com> wrote:

> Would you mind grabbing the lint errors when you get a moment?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/19729#issuecomment-69944132>.
>

jacksontj committed:
Pylint cleanup

kiorky commented:
Uhm, @cachedout @jacksontj i'm -1 on running on a thread, why do not call directly the function ?
We are in sync mode anyway here, no gain to spawn a thread isnt it ?
Moreover, not running it in a thread will reallow to use pdb :-) !

kiorky commented:
Uhm there may be a race as if we are not threaded, the event would be thrown, but only at the end and not from the execution flow blocking an eventual runner playing with the even bus to collaborate on other events isnt it ?

jacksontj commented:
We are spawning a thread because we have changed how the print stuff works. The major change @cachedout  was trying to add was that you could print out progress of a runner while its executing. To do so we fire events and then print out the events in the main thread. We could do it not in a thread but we'd have to print when we fire the event in the runner-- which is completely different from the daemonized runner.

As for a race condition, there shouldn't be. If you see the "new" event, then the event listener is already subscribed.

kiorky commented:
Yep; our comments just cross themselves, we are on the same line.

Maybe we can add some sort of switch to OR not to thread. My only concern here is to have pdb a

kiorky commented:
Yep; our comments just cross themselves, we are on the same line.

Maybe we can add some sort of switch to OR not to thread. My only concern here is to have pdb available when creating new runners, and having some corner cases

jacksontj commented:
Possible... I'll have to take a look tomorrow.. We might just be able to overload the event thing in the runner to print as well as fire an event, then we won't need threading... I'll check it out. I'm trying to strike a balance between them being the same (as the old impl) and getting the new functionality without diverging too much.

jacksontj commented:
@cachedout Actually the more I think about this feature, I think we can get what the intended behavior was with less work. If I understand correctly, the intent was to allow runners to report status/progress as they are running. They could do this before if you were running them using "salt-run" by using print. Was the intent to make this exposed outside the master? If so I think we can revert most of the code change we've made and just change the event fire to check for a specific opt that says "print me as well" and just set that when you run it in the foreground.

If there was some other use-case I'm missing, let me know. But I think that would clear up a lot of this magic we've put in here.

cachedout commented:
Hi @jacksontj 

The original intention here was that runners could broadcast what were formerly just print calls across the event bus so that any event-bus observer could see the progress of the runner as it went along. If we can simplify things and still accomplish that, I'm definitely all for it. :]

jacksontj commented:
Yea, I think with that specific intent in mind we can do this in a lot less magical way... I'll try to put something together tomorrow ;)

cachedout commented:
Go Go Jenkins

kiorky commented:
What's the status on this now ? @jacksontj did you had some time to think about the new change you had ?

With the current code, i have always intermittent non returns from runners :(

kiorky commented:
it seems stuck in https://github.com/saltstack/salt/blob/develop/salt/utils/event.py#L360

jacksontj commented:
Sorry about the delay, been so busy with everything else. I'm hoping to get time for this tomorrow, but definitely this week.

kiorky commented:
With those changesets, i get an error on the console (just a bit more output); but the command is always hanging, and the control C is at the same point, on the infinite loop
```
^C[ERROR   ] Traceback (most recent call last):
  File "/srv/mastersalt/makina-states/src/salt/salt/scripts.py", line 252, in salt_run
    client.run()
  File "/srv/mastersalt/makina-states/src/salt/salt/cli/run.py", line 54, in run
    runner.run()
  File "/srv/mastersalt/makina-states/src/salt/salt/runner.py", line 169, in run
    for suffix, event in self.get_async_returns(async_pub['tag'], event=self.event):
  File "/srv/mastersalt/makina-states/src/salt/salt/client/mixins.py", line 367, in get_async_returns
    raw = event.get_event(timeout, tag=tag, full=True)
  File "/srv/mastersalt/makina-states/src/salt/salt/utils/event.py", line 430, in get_event
    ret = self._get_event(wait, tag, pending_tags, match_func)
  File "/srv/mastersalt/makina-states/src/salt/salt/utils/event.py", line 360, in _get_event
    socks = dict(self.poller.poll(wait * 1000))
  File "/srv/mastersalt/makina-states/eggs/pyzmq-14.4.1-py2.7-linux-x86_64.egg/zmq/sugar/poll.py", line 101, in poll
    return zmq_poll(self.sockets, timeout=timeout)
  File "zmq/backend/cython/_poll.pyx", line 115, in zmq.backend.cython._poll.zmq_poll (zmq/backend/cython/_poll.c:1589)
  File "zmq/backend/cython/checkrc.pxd", line 11, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/_poll.c:1793)
    PyErr_CheckSignals()
  File "/srv/mastersalt/makina-states/eggs/async-0.6.1-py2.7-linux-x86_64.egg/async/__init__.py", line 21, in thread_interrupt_handler
    prev_handler(signum, frame)
KeyboardInterrupt

[ERROR   ] An un-handled exception was caught by salt's global exception handler:
KeyboardInterrupt: 
Traceback (most recent call last):
  File "/srv/mastersalt/makina-states/bin/salt-run", line 66, in <module>
    sys.exit(salt.scripts.salt_run())
  File "/srv/mastersalt/makina-states/src/salt/salt/scripts.py", line 262, in salt_run
    hardcrash, trace=trace)
  File "/srv/mastersalt/makina-states/src/salt/salt/scripts.py", line 39, in _handle_interrupt
    raise original_exc
KeyboardInterrupt
Traceback (most recent call last):
  File "/srv/mastersalt/makina-states/bin/salt-run", line 66, in <module>
    sys.exit(salt.scripts.salt_run())
  File "/srv/mastersalt/makina-states/src/salt/salt/scripts.py", line 262, in salt_run
    hardcrash, trace=trace)
  File "/srv/mastersalt/makina-states/src/salt/salt/scripts.py", line 39, in _handle_interrupt
    raise original_exc
KeyboardInterrupt

```

kiorky commented:
I'll try to have a look tomorrow too, :)

kiorky commented:
The runner outputs a lot to the console and also a lot of events with attached data, so that's why im thinking on somewhere that doesent have the time either to pull from or push to something (event, data, whatever)

kiorky commented:
i fixed my problems with jacksontj/salt/pull/1

jacksontj committed:
Rework the runner CLI to not use threads

No reason to go through all that work and introduce debugging issues. This way we can still fire events using __jid_event__ but if you just "print" it will do so without trying to monkey patch print

jacksontj committed:
Revert "Fix two problems with runners"

This reverts commit 389fb7e030b3cb340cec6c6d4975d95887819c13.

Conflicts:
	salt/client/mixins.py

jacksontj commented:
Also included a revert of #389fb7e as it shouldn't be required anymore with this last re-work

jacksontj committed:
Remove unused code

jacksontj committed:
Nicer error handling for runner execution

jacksontj commented:
@kiorky  #5d07de5 includes both of your patches.

jacksontj committed:
Pylint cleanup

jacksontj commented:
@cachedout Once this passes I believe we are complete. Then we'll just need to backport this to 2014.7 and then we are good :)

cachedout commented:
Sounds great! We'll try and get this merged today. Thanks again for your great work here.

kiorky commented:
@jacksontj my patches were not applied correctly im sending you a rework

kiorky commented:
Hold on merge, i will try the rework.
(Im going to update the status in the next 2 hours.)

cachedout commented:
No problem. Just let us know when you're ready. :]

On Fri, Jan 23, 2015 at 10:53 AM, Mathieu Le Marec - Pasquet <
notifications@github.com> wrote:

> Hold on merge, i will try the rework.
> (Im going to update the status in the next 2 hours.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/19729#issuecomment-71233999>.
>

kiorky commented:
Seems not working at all
```
Traceback (most recent call last):
  File "/srv/mastersalt/makina-states/src/salt/salt/client/mixins.py", line 232, in low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/srv/mastersalt/makina-states/mc_states/runners/mc_cloud_saltify.py", line 192, in orchestrate
    _s['mc_api.time_log']('start', func_name, skip=skip, only=only)
  File "/srv/mastersalt/makina-states/mc_states/runners/mc_api.py", line 64, in time_log
    msg, event_category=lifecycle_id, fun=fun, *args, **extra_data)
  File "/srv/mastersalt/makina-states/mc_states/runners/mc_api.py", line 55, in cloud_event
    __jid_event__.fire_event(payload, tag)
  File "/srv/mastersalt/makina-states/src/salt/salt/utils/event.py", line 618, in fire_event
    self.print_func(tag, data)
  File "/srv/mastersalt/makina-states/src/salt/salt/client/mixins.py", line 341, in print_async_event
    event.pop('_stamp')  # remove the timestamp before printing
KeyError: '_stamp'


```

kiorky commented:
is that me or the pull request has something unrelated to runners (the key stuff ??? )
@jacksontj

kiorky commented:
nvm, ... sorry for the noise

kiorky commented:
I mean, for the keys stuff, i have still the trace, and i'm diffing the whole between the code delta between @jacksontj fork, salt/develop and our fork...

kiorky commented:
Ok, so no code delta, the trace is legit and has to be investigated :/

kiorky commented:
im using a lot events in our custom runners, something has changed and require _stamp in the payload, im looking why there is a change here and whatever fix to do (either to salt or our custom implem)

kiorky commented on salt/client/mixins.py:
```diff
@@ -341,40 +343,3 @@ def print_async_event(self, suffix, event):
             event.pop('_stamp')  # remove the timestamp before printing
```
problem is here:
1. i dont think we should pop stuff from the payload as it will disappear from the mutable dict
2. if we dont want the _stamp in the ouput, we then have to do elseway (without deepcopy for obious perf reasons)

kiorky replied:
In the case of our custom event, it got printed out without having a chance to be completed with the _stamp, causing the crash at the pop statement

kiorky replied:
we are thinking about keeping the _stamp in the printed data, and not doing the pop with @cachedout.

What's your point @jacksontj ?

kiorky commented:
@jacksontj i added a commit not to filter _stamp to my PR on your fork.

jacksontj committed:
Conditionally pop _stamp

jacksontj committed:
Keep the stamp

jacksontj commented:
A rebase on newer develop

jacksontj committed:
Don't *pop* the outputter, just pass opts of None

kiorky commented:
Again, hold on / do stop any merge on this, i am fixing 2 new oustanding problems:
* jobs returns are not going to master jobs cache (no mastercall is throwed)
* event jid printout will not work as we do not use an enought powerfull loglevel (by default)

Wait for my PR on @jacksontj for this (hoping this once, he will merge it correctly)

kiorky committed:
runner: Print out correctly jid, both in sync/async

jacksontj commented:
@kiorky I don't see any open PRs?

kiorky commented:
im still working on it

kiorky committed:
runner/master Store jobs in cache, & factorization

kiorky commented:
@jacksontj right now, we have jobs getting back to cache dir, but we have a side effect somewhere which remove the jobs from the cache or overwrites it.
Indeed, if i use lookup_jid, it returns data from something like 30 seconds but nothing after.

kiorky committed:
runner return: also add tgt to the payload

kiorky committed:
runner: Fix job store

kiorky committed:
runner: unclutter the output from mminion loading

kiorky commented:
So, the PR i made to test with @cachedout is proprosed for review to @jacksontj which can merge my commits in that PR for it to be merged afterall.

jacksontj commented:
@kiorky historically we didn't save the runner returns in the job cache at all. I don't think we *shouldnt* but for this PR i'm more concerned about getting compatibility working for the 2015.2 release which is coming up soon, so I'd almost rather not include that code change just since its quite a bit of new ;/.

I'm also wondering if we should even print out the events on the CLI at all-- since we haven't historically. Maybe we just have the runners go back to printing just the returns that they did before and just fire events? I just want to get this all closed out ASAP so we don't release 2015.2 with backwards incompatible changes.

cc @cachedout

kiorky committed:
Removing useless cruft

kiorky committed:
Rework a bit the runner completion log

jacksontj committed:
Misc style cleanup

jacksontj committed:
Already prints out the output if run in sync mode, and async already has this in the ret event

jacksontj commented:
@cachedout Assuming these tests pass I think we are happy. I still dislike how there is no mechanism to get the simple output we used to have... besides that we are happy with it

jacksontj commented:
@rallytime Both tests "aborted" for some reason :/

rallytime commented:
Test vms didn't bootstrap. Let's try again.

rallytime commented:
Go Go Jenkins!

rallytime commented:
Ok, it's a little hard to tell on this interface, but we have indeed started a new run.

rallytime commented:
CentOS 7: https://jenkins.saltstack.com/job/salt-pr-rs-cent7-n/786/
Ubuntun 14: https://jenkins.saltstack.com/job/salt-pr-rs-ubuntu14.04-n/786/
Pylint: https://jenkins.saltstack.com/job/salt-pr-lint-n/1219/

jacksontj commented:
@rallytime those seem to have grey circles on them? Does that mean aborted?

rallytime commented:
Usually they mean something timed out, but they can also mean aborted sometimes. We're having some bad luck here, aren't we?

rallytime commented:
Go Go Jenkins!

rallytime commented:
Ok, on this pull request the tests aren't running because we're stacktracing due the the changes in `output/__intit__.py`:
```
Traceback (most recent call last):
03:07:17   File "/testing/tests/runtests.py", line 411, in <module>
03:07:17     main()
03:07:17   File "/testing/tests/runtests.py", line 396, in main
03:07:17     status = parser.run_integration_tests()
03:07:17   File "/testing/tests/runtests.py", line 316, in run_integration_tests
03:07:17     with TestDaemon(self):
03:07:17   File "/testing/tests/integration/__init__.py", line 219, in __enter__
03:07:17     salt.output.display_output(grains, 'grains', minion_opts)
03:07:17   File "/testing/salt/output/__init__.py", line 86, in display_output
03:07:17     display_data = try_printout(data, out, opts)
03:07:17   File "/testing/salt/output/__init__.py", line 45, in try_printout
03:07:17     log.error('Nested output failed: ', exec_info=True)
03:07:17   File "/usr/lib/python2.7/logging/__init__.py", line 1178, in error
03:07:17     self._log(ERROR, msg, args, **kwargs)
03:07:17 TypeError: _log() got an unexpected keyword argument 'exec_info'
```

rallytime commented:
Ok...so the weird thing is when I aborted the CentOS 7 VM because it was just spinning forever instead of sending test results back to the console output on jenkins, we actually got the test results back of tests that were running. I haven't been following this thread closely, but did something change here that isn't working with the test runner and test output?

ping @s0undt3ch

jacksontj committed:
Fix typo from 37619f0e

jacksontj commented:
@rallytime That stacktrace is actually from 37619f0e @thatch45 typo-d

jacksontj commented:
@rallytime these 3 failures seem to be related to changes with immutabletypes, are you sure those 3 aren't seen on develop?

rallytime commented:
@jacksontj The wheel.key failures are not present on develop, but the file_test one is. So the two pylint errors and the file_test failure are not applicable here, but the key ones are. I'll have to look closer here to see why that is failing.

jacksontj committed:
Pass opts not an ImmutableDict of opts

jacksontj committed:
Lazily load masterminion and keep it around.

jacksontj commented:
@rallytime Looks like someone added some frozen types, and the harness was passing in one when it shouldn't have ;)

rallytime commented:
Ah, yes. Thanks for looking at that!

rallytime commented:
@thatch45 All test failures are unrelated! :D

cachedout commented:
As far as I'm concerned, this is ready to be merged in. Any last changes @jacksontj or @kiorky? If not, I think we'll get this in today. :]

kiorky commented:
i checked in the changeset about localmastersaltminion, but did not tested it, althought it seems harmless.

kiorky commented:
(so yes, for me it is now mergeable)
